[DOCS] Timelion Getting Started Guide

alexfrancoeur opened the pull request:
This is the first iteration of the Timelion Getting Started Guide. I have added a number of images and split up the documentation into 6 different getting started areas. 

* Create a time series chart with multiple expressions
* Customize the visualization type and formatting options
* Use mathematical functions to manipulate your data
* Track trends and use conditional logic to format outliers
* Add Timelion visualizations to a dashboard
* Reference inline help

I'd like to review the content further to make sure the expressions make sense and the flow is intuitive.

alexfrancoeur commented:
@tbragin ran through this again and updated some minor things. Added you as a reviewer, let me know if you have any feedback.

@rashidkpc I added you as a reviewer as well. Would love your thoughts on the content, flow and most importantly the Timelion expressions. I realize there are some arguments that might not be necessary in some cases (timefield for example) but I thought it was worth pointing out how to define. Any feedback here would be appreciated.

tbragin commented:
Scheduled a live sync on this. Feedback discussed so far this morning:
* Long expressions go off page. It's nice to copy/paste them, but not readable. @debadair advice here? How can we have our cake and eat it too? :)
* Adjust tone in places for consistency.
* Broken links (Alex explained) will work online when merged -- they are relative links. 
* Explain the change in each new expression - add a sentence, link to reference? (to discuss)
* Stick to using named parameters for clarity
* Explain if in more detail, point to blog https://www.elastic.co/blog/timeseries-if-then-else-with-timelion
* At some point, drop timefield so it doesn't get repeated everywhere? Other ways to shorten the expressions? Discussed variables, but decided not to use for now. 
* Inline code formatting instead of bolding
* Point to additional Timelion resources, such as meetup videos, in intro or separate section?
https://www.elastic.co/elasticon/conf/2017/sf/timelion-magic-math-and-everything-in-the-middle
https://www.elastic.co/videos/timelion-plugin-for-kibana-enables-times-series-paris-meetup
https://www.elastic.co/videos/using-kibana-and-timelion-to-analyze-earthquake-data

alexfrancoeur commented:
Latest push has an adjusted tone, function / argument explanations, link to blog and code formatting instead of bolding. Would like to discuss dropping the timefield and where to point to additional resources before adding them in.

alexfrancoeur commented:
Made the following updates based on yesterday's discussion:
- Added {nbsp} after images to format better
- Made getting started a chapter with sub-sections and added reference to videos
- Removed the "author's we"

alexfrancoeur commented:
Also, based on other sections that I've built locally - the "on this page" section in the top right ends up becoming a subsection of getting started when merged that is expandable / collapsable.

<img width="629" alt="screen shot 2017-04-17 at 7 12 31 am" src="https://cloud.githubusercontent.com/assets/25181823/25087482/48b8b5dc-233d-11e7-9308-448f65e771b8.png">

bohyun-e commented on docs/timelion/getting-started/timelion-conditional.asciidoc:
```diff
@@ -0,0 +1,62 @@
+[[timelion-conditional]]
+=== Using conditional logic and tracking trends
+
+In this section you will learn how to modify a series with conditional logic and trend your data with a moving average. This is helpful to easily distinguish outliers and detect trends over time.
```
To me, 'a series' is a bit ambiguous. Can we be more descriptive like 'time-series data'?

bohyun-e replied:
s/"distinguish outliers and detect trends over time"/"detect outliers and patterns over time"

bohyun-e commented on docs/timelion/getting-started/timelion-conditional.asciidoc:
```diff
@@ -0,0 +1,62 @@
+image::images/timelion-conditional01.png[]
+{nbsp}
+
+Let’s create two thresholds to keep an eye on the amount of memory your machine is using. For the purposes of this tutorial, your warning threshold will be 12.5GB and your severe threshold will be 15GB. When the maximum amount of memory being utilized exceeds either of these thresholds, the series will be re-styled accordingly.
```
s/"the amount of memory your machine is using"/"the amount of used memory"

bohyun-e replied:
Let's use consistent terminology. s/"memory being utilized"/"used memory"

bohyun-e replied:
What does it mean by "the series will be re-styled"? Do we mean that they will have different colors?

alexfrancoeur replied:
Colors & line width, but the line width is an extra touch towards the end of this section. I'd be fine with replacing "the series will be re-styled accordingly" with "the series will be colored accordingly"

bohyun-e commented on docs/timelion/getting-started/timelion-conditional.asciidoc:
```diff
@@ -0,0 +1,62 @@
+
+NOTE: If the threshold values are too high or low for your machine, please adjust accordingly.
+
+To configure these two threshold values, you can utilize Timelions conditional logic. In this tutorial you will use `if()` to compare each point to a number, adjust the styling if the comparison is `true` and use the default styling if the comparison is `false`. Timelion offers six operator values for comparison.
```
s/"Timelions"/"Timelion's"

bohyun-e replied:
if the comparison is `true` and use the default styling if the comparison is `false` => if the condition evaluates to `true` and use the default styling if the condition evaluates to`false`

bohyun-e replied:
Add "the following" between 'Timelion offers' and 'six operator'.

alexfrancoeur replied:
++

bohyun-e commented on docs/timelion/getting-started/timelion-conditional.asciidoc:
```diff
@@ -0,0 +1,62 @@
+
+For additional information on Timelions conditional capabilities, check out the blog post https://www.elastic.co/blog/timeseries-if-then-else-with-timelion[I have but one .condition()].
+
+Now that you have thresholds defined to easily identify outliers, let’s create a new series to determine what the trend really is. Timelions `mvavg()` function allows you to calculate the moving average over a given window. This is especially helpful for noisey time series. For this tutorial, you will use `.mvavg(10)` to create a moving average with a window of 10 data points. Use the following e ...
```
s/"Timelions"/"Timelion's"

bohyun-e replied:
we use 'memory consumption' here, but for consistent language, should we say 'memory usage'?

bohyun-e commented on docs/timelion/getting-started/timelion-create.asciidoc:
```diff
@@ -0,0 +1,37 @@
+
+This tutorial will be using the time series data from https://www.elastic.co/guide/en/beats/metricbeat/current/index.html[Metricbeat] to walk you through a number of functions that Timelion offers. To get started, download Metricbeat and follow the https://www.elastic.co/downloads/beats/metricbeat[instructions here] to start ingesting the data locally.
+
+The first visualization you will create compares the average percentage of CPU time spent in user space to the previous hours. In order to create this visualization, we’ll need to create two Timelion expressions. One with the current average of `system.cpu.user.pct` and another with the average from the previous hour.
```
>The first visualization you will create compares the average percentage of CPU time spent in user space to the previous hours.

This sentence is not clear. What's the bucket interval for the average percentage of CPU time spent in user space that is being compared to that of the previous hours? 

>One with the current average of `system.cpu.user.pct` and another with the average from the previous hour.

This sentence says "previous hour". I'm not sure which time bucket we are comparing.

alexfrancoeur replied:
@bohyun-e how does this sound? Any better?

> The first visualization you'll create will compare the real-time percentage of CPU time spent in user space to the results offset by one hour. In order to create this visualization, we’ll need to create two Timelion expressions. One with the real-time average of `system.cpu.user.pct` and another with the average offset by one hour.

bohyun-e commented on docs/timelion/getting-started/timelion-create.asciidoc:
```diff
@@ -0,0 +1,37 @@
+image::images/timelion-create01.png[]
+{nbsp}
+
+Now you need to add a another series with data from the previous hour for comparison. To do so, you'll have to add an `offset` arguement to the `.es()` function. `offset` will offset the series retrieval by a date expression. For this example, you'll want to look at the previous hour and will be using the date expression `-1h`. Using a comma to separate the two series, enter the following express ...
```
drop "a" before "another series..."

alexfrancoeur replied:
good catch :-)

bohyun-e commented on docs/timelion/getting-started/timelion-customize.asciidoc:
```diff
@@ -0,0 +1,53 @@
+image::images/timelion-customize02.png[]
+{nbsp}
+
+Let’s color these series so that the current hour pops a bit more than the last hour. The `color()` function can be used to change the color of any series and accepts standard color names, hexadecimal value or a color schema for grouped series. For this example you will use `.color(gray)` for the last hour and `.color(#1E90FF)` for the current hour. Enter the following expression into the Timelio ...
```
To stay consistent with the language you've been using, 
s/"the current hour"/"the current hour series"

bohyun-e replied:
hexadecimal values

bohyun-e commented on docs/timelion/getting-started/timelion-math.asciidoc:
```diff
@@ -0,0 +1,61 @@
+image::images/timelion-math02.png[]
+{nbsp}
+
+Now that you have the maximum value for the inbound traffic, let’s calculate the derivative to chart the change over time. The `derivative()` function is used to plot the change in values over time. This can be easily done by appending the `.derivative()` to the end of an expression. Use the following expression to update your visualization:
```
>Now that you have the maximum value for the inbound traffic, let’s calculate the derivative to chart the change over time.

This sentence isn't clear from just reading. I understand by seeing the expression below, however. I might suggest something like "We are interested in plotting the change of X over time, so we will use the derivate function,"

alexfrancoeur replied:
I agree, though I've been told that we shouldn't using "we" and "we will" in the tutorial. How's this?

>Monitoring network traffic is much more valuable when plotting the rate of change. The `derivative()` function is used do just that - plot the change in values over time. This can be easily done by appending the `.derivative()` to the end of an expression.

alexfrancoeur commented:
@bohyun-e made some modifications based on your suggestions. Thanks again for taking the time to review. Let me know your thoughts.

alexfrancoeur commented:
@rashidkpc @tbragin @debadair I'd like to get this guide in around the same time as 5.4. Any feedback next week would be greatly appreciated.

tbragin commented:
@alexfrancoeur LGTM
@debadair @rashidkpc @bohyun-e Any comments before we merge?

rashidkpc commented:
I really like both the voice and the level of detail in the updates. It a great walkthrough, with helpful notes and take away along the way. 👍  from me.
